src/lib/src/common/geometry/interpolate.cpp

```diff
@@ -0,0 +1,155 @@
+#include <array>
+#include <cmath>
+#include <optional>
+#include <vector>
+
+#include <pybind11/stl.h>
+
+#include <xtgeo/geometry.hpp>
+#include <xtgeo/logging.hpp>
+#include <xtgeo/xtgeo.h>
+
+namespace xtgeo::geometry {
+
+// local helper function to interpolate Z in a triangle
+static double
+interpolate_z_triangles(const double x,
+                        const double y,
+                        const std::array<double, 3> &p1,
+                        const std::array<double, 3> &p2,
+                        const std::array<double, 3> &p3,
+                        const std::array<double, 3> &p4)
+{
+    // Convert 3D points to 2D points for area calculation, generic
+    std::array<double, 2> p1_2d = { p1[0], p1[1] };
+    std::array<double, 2> p2_2d = { p2[0], p2[1] };
+    std::array<double, 2> p3_2d = { p3[0], p3[1] };
+    std::array<double, 2> p4_2d = { p4[0], p4[1] };
+    std::array<double, 2> p = { x, y };
+
+    // Compute areas of the triangles
+    double area1 = triangle_area(p, p2_2d, p3_2d);
+    double area2 = triangle_area(p1_2d, p, p3_2d);
+    double area3 = triangle_area(p1_2d, p2_2d, p);
+    double total_area = triangle_area(p1_2d, p2_2d, p3_2d);
+
+    double z_estimated = std::numeric_limits<double>::quiet_NaN();
+
+    // Check if the point is in the triangle
+    if (std::abs(total_area - (area1 + area2 + area3)) < 1e-6) {
+        double w1 = area1 / total_area;
+        double w2 = area2 / total_area;
+        double w3 = area3 / total_area;
+        z_estimated = w1 * p1[2] + w2 * p2[2] + w3 * p3[2];
+    }
+    // check that z_estimated is not NaN
+    if (!std::isnan(z_estimated)) {
+        return z_estimated;
+    }
+    // if still NaN, check the other triangle
+    area1 = triangle_area(p, p3_2d, p4_2d);
+    area2 = triangle_area(p1_2d, p, p4_2d);
+    area3 = triangle_area(p1_2d, p3_2d, p);
+    total_area = triangle_area(p1_2d, p3_2d, p4_2d);
+
+    // Check if the point is in the triangle
+    if (std::abs(total_area - (area1 + area2 + area3)) < 1e-6) {
```

Comment:
Should 1e-6 be a constant threshold for other use cases as well?

Reply:
Convert this to a constant names EPSILON